cdc/processor/sinkmanager/worker_impl.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sinkmanager
+
+import (
+	"context"
+
+	"github.com/pingcap/tiflow/cdc/model"
+	"github.com/pingcap/tiflow/cdc/redo"
+	"github.com/pingcap/tiflow/pkg/sorter"
+)
+
+// Assert that workerImpl implements worker.
+var _ worker = (*workerImpl)(nil)
+
+type workerImpl struct {
+	parentCtx   context.Context
+	redoManager redo.LogManager
+	sortEngine  sorter.EventSortEngine
+	memQuota    memQuota
+	splitTxn    bool
+	batchSize   uint64
+	closedChan  chan struct{}
+}
+
+func (w *workerImpl) run(taskChan <-chan *tableSinkTask) error {
+	for {
+		select {
+		case <-w.closedChan:
+			return nil
+		case t := <-taskChan:
+			// First time to run the task, we have initialized memory quota for the table.
+			availableMem := defaultMemoryUsage
+			events := make([]*model.PolymorphicEvent, 0, 1024)
+			var lastPos sorter.Position
+			batchCount := 0
+			iter := w.sortEngine.FetchByTable(t.tableID, t.lowerBound, t.upperBound)
+			for {
+				e, pos, err := iter.Next()
+				if err != nil {
+					return err
+				}
+				// There is no more data.
+				if e == nil {
+					break
+				}
+				for availableMem-e.Row.ApproximateBytes() < 0 {
+					w.memQuota.ForceAcquire()
+					availableMem += defaultMemoryUsage
+				}
+				availableMem -= e.Row.ApproximateBytes()
+				events = append(events, e)
+				// We meet a finished transaction.
+				if pos.Valid() {
+					lastPos = pos
+					if w.splitTxn {
+						w.memQuota.ResetBatchID(t.tableID)
+					}
+					// Always emit the events to the sink.
+					// Whatever splitTxn is true or false, we should emit the events to the sink as soon as possible.
+					err := w.emitEventsToTableSink(events, t, e.CRTs)
```

Comment:
If there are lots of small transactions, it can call `memQuota.Record` too much times. Maybe it's a problem.
How about just writeEvents into the sink, but update resolved ts later? For example, we can update resolved ts per 256K.

Reply:
So we can call `memQuota.Record` with `sink.UpdateResolvedTs`, to reduce the frequency.

Reply:
Do we send too infrequently if it causes delays? For example, it can't reach 256k in a short time? Do we have to introduce a timeout for this?

Reply:
Timeout seems unnecessary. We always do a flush when this task is finished.

Reply:
I see. I'll try it.

Reply:
Added. Please check out https://github.com/pingcap/tiflow/pull/7388/commits/09151209d27f18bfa0b7b6bfd97d8966b5081909